File: src/Python/pipeline_services/core/registry.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict

from .pipeline import StepAction, UnaryOperator


@dataclass
class PipelineRegistry:
    unary_actions: Dict[str, UnaryOperator] = field(default_factory=dict)
    step_actions: Dict[str, StepAction] = field(default_factory=dict)

    def register_unary(self, name: str, action: UnaryOperator) -> None:
        self.unary_actions[name] = action

    def register_action(self, name: str, action: StepAction) -> None:
        self.step_actions[name] = action

    def has_unary(self, name: str) -> bool:
        return name in self.unary_actions

    def has_action(self, name: str) -> bool:
        return name in self.step_actions

    def get_unary(self, name: str) -> UnaryOperator:
        if name in self.unary_actions:
            return self.unary_actions[name]
        raise ValueError("Unknown unary action: " + name)

    def get_action(self, name: str) -> StepAction:
        if name in self.step_actions:
            return self.step_actions[name]
        raise ValueError("Unknown step action: " + name)

File: src/typescript/src/pipeline_services/core/registry.ts
import { StepAction, UnaryOperator } from "./pipeline.js";

export class PipelineRegistry {
  private unary_actions: Map<string, UnaryOperator>;
  private step_actions: Map<string, StepAction>;

  constructor() {
    this.unary_actions = new Map<string, UnaryOperator>();
    this.step_actions = new Map<string, StepAction>();
  }

  register_unary(name: string, action: UnaryOperator): void {
    this.unary_actions.set(name, action);
  }

  register_action(name: string, action: StepAction): void {
    this.step_actions.set(name, action);
  }

  has_unary(name: string): boolean {
    return this.unary_actions.has(name);
  }

  has_action(name: string): boolean {
    return this.step_actions.has(name);
  }

  get_unary(name: string): UnaryOperator {
    const action = this.unary_actions.get(name);
    if (action != null) {
      return action;
    }
    throw new Error("Unknown unary action: " + name);
  }

  get_action(name: string): StepAction {
    const action = this.step_actions.get(name);
    if (action != null) {
      return action;
    }
    throw new Error("Unknown step action: " + name);
  }
}

